d5005/hardware/common/build/ddr_board_hw.tcl

```diff
@@ -0,0 +1,250 @@
+set_parameter_property NUMBER_OF_MEMORY_BANKS DISPLAY_NAME "Number of Memory Banks"
+set_parameter_property NUMBER_OF_MEMORY_BANKS AFFECTS_ELABORATION true
+
+add_parameter MEMORY_BANK_ADDRESS_WIDTH INTEGER 32
+set_parameter_property MEMORY_BANK_ADDRESS_WIDTH DEFAULT_VALUE 32
+set_parameter_property MEMORY_BANK_ADDRESS_WIDTH DISPLAY_NAME "Memory Bank Address Width"
+set_parameter_property MEMORY_BANK_ADDRESS_WIDTH AFFECTS_ELABORATION true
+
+add_parameter SNOOP_PORT_ENABLE BOOLEAN false
+set_parameter_property SNOOP_PORT_ENABLE DEFAULT_VALUE false
+set_parameter_property SNOOP_PORT_ENABLE DISPLAY_NAME "Enable Snoop Port"
+set_parameter_property SNOOP_PORT_ENABLE AFFECTS_ELABORATION true
+
+add_parameter MBD_TO_MEMORY_PIPE_STAGES INTEGER 0
+set_parameter_property MBD_TO_MEMORY_PIPE_STAGES DEFAULT_VALUE 0
+set_parameter_property MBD_TO_MEMORY_PIPE_STAGES DISPLAY_NAME "MBD to Memory Pipeline Stages"
+set_parameter_property MBD_TO_MEMORY_PIPE_STAGES AFFECTS_ELABORATION true
+# | 
+# +-----------------------------------
+
+proc compose { } {
+  # Get parameters
+  set number_of_memory_banks    [ get_parameter_value NUMBER_OF_MEMORY_BANKS ]
+  set memory_bank_address_width [ get_parameter_value MEMORY_BANK_ADDRESS_WIDTH ]
+  set snoop_port_enable         [ get_parameter_value SNOOP_PORT_ENABLE ]
+  set mbd_to_memory_pipe_stages [ get_parameter_value MBD_TO_MEMORY_PIPE_STAGES ]
+
+  # Compute parameters
+  set log2_num_banks [ expr log($number_of_memory_banks) / log(2) ]
+  set total_address_width [ expr $memory_bank_address_width + $log2_num_banks ]
+
+  # Instances and instance parameters
+  add_instance host_clk altera_clock_bridge 19.2.0
+  set_instance_parameter_value host_clk {EXPLICIT_CLOCK_RATE} {0.0}
+  set_instance_parameter_value host_clk {NUM_CLOCK_OUTPUTS} {1}
+  
+  add_instance kernel_clk altera_clock_bridge 19.2.0
+  set_instance_parameter_value kernel_clk {EXPLICIT_CLOCK_RATE} {0.0}
+  set_instance_parameter_value kernel_clk {NUM_CLOCK_OUTPUTS} {1}
+
+  for { set i 0} { $i < $number_of_memory_banks } {incr i} {
+    add_instance ddr_clk$i altera_clock_bridge 19.2.0
+    set_instance_parameter_value ddr_clk$i {EXPLICIT_CLOCK_RATE} {0.0}
+    set_instance_parameter_value ddr_clk$i {NUM_CLOCK_OUTPUTS} {1}
+  }
+
+  set global_reset_outputs [expr $number_of_memory_banks + 1]
+  add_instance global_reset altera_reset_bridge 19.2.0
+  set_instance_parameter_value global_reset {ACTIVE_LOW_RESET} {0}
+  set_instance_parameter_value global_reset {SYNCHRONOUS_EDGES} {both}
+  set_instance_parameter_value global_reset {NUM_RESET_OUTPUTS} $global_reset_outputs
+
+  add_instance kernel_reset altera_reset_bridge 19.2.0
+  set_instance_parameter_value kernel_reset {ACTIVE_LOW_RESET} {0}
+  set_instance_parameter_value kernel_reset {SYNCHRONOUS_EDGES} {deassert}
+  set_instance_parameter_value kernel_reset {NUM_RESET_OUTPUTS} {1}
+
+  for { set i 0} { $i < $number_of_memory_banks } {incr i} {
+    add_instance ddr_channel$i ddr_channel 23.2
```

Comment:
should these version numbers be parameters/defines? There are earlier 19.2.0 - can they not be updated to 23.2 as well?

Reply:
Unfortunately no. These are PD IPs and when you check in the original PD system you will find that those are version 19.2.0 (I am guessing when this was updated last). The others with 23.2 are our own IP which we defined as being version 23.2.

Reply:
I believe if you would change to user newer versions of those PD IPs it will automatically revert back to the latest ones, but will result in warnings everywhere.